README.md
# LOMAR

An R package to deal with point sets from single molecule localization microscopy.
This package provides three sets of functionalities:
  - data input: read SMLM data as point sets either from csv files or from TIFF images
  - registration: point sets registration using different algorithms
  - topological data analysis: compute similarity between point sets using persistent homology

# Installation

The package is [available on CRAN](https://cran.r-project.org/package=LOMAR).

![Download counts](https://cranlogs.r-pkg.org/badges/grand-total/LOMAR) (from the RStudio mirror)

``` R
install.packages("LOMAR")
```

To install the development version of this package, run (from within R):

``` R
library(devtools)
install_git('https://git.embl.de/heriche/lomar')
```

This package depends on these other packages:
  * data.table
  * FNN
  * foreach
  * parallel
  * doParallel
  * proxy
  * reshape2
  * pracma
  * transport
  * RANN
  * ff
  * aws
  * dbscan
  * EBImage (from Bioconductor)
  


NEWS.md
# LOMAR 0.2.1

## Changes
  * Fixed C++ deprecation warnings about std::binary_function and std::unary_function

# LOMAR 0.2.0

## Changes
  * Removed dependency on archived package TDA by importing the necessary code. The default TDA library used is now Dionysus.
  * Changed license to GPL-3 to be compatible with code from the TDA package.
  
# LOMAR 0.1.0
  * Initial release on CRAN


tests/testthat/test-TDA.R
set.seed(20211123)

data.file <- system.file("test_data", "simulated_NUP107_data.csv", package="LOMAR")
locs <- locs_from_csv(data.file)
PS <- point_sets_from_locs(locs = locs, keep.locprec = FALSE)

test_that("Persistence diagrams can be computed", {
  Diag <- get_persistence_diagrams(PS, maxdimension = 1, maxscale = 100)
  expect_equal(length(Diag), length(PS))
  expect_equal(length(which(Diag[[3]][, "dimension"] == 1)), 8)
})

test_that("Sliced Wasserstein distance matrix can be computed", {
  Diag <- get_persistence_diagrams(PS[1:20], maxdimension = 1, maxscale = 100)
  sWD <- get_kernel_matrix(Diag, method = "sWd", M = 10, return.dist = TRUE, ncpu = 2)
  dt <- determinant(sWD)
  expect_equal(as.numeric(floor(dt$modulus)), 106)
})

tests/testthat/test-utils.R
set.seed(20211123)

data.file <- system.file("test_data", "simulated_NUP107_data.csv", package="LOMAR")
locs <- locs_from_csv(data.file)
PS <- point_sets_from_locs(locs = locs, keep.locprec = TRUE)

test_that("Cropping point set works", {
  pts <- crop_point_set(PS[[2]], c(60, 60, 120))
  expect_equal(nrow(pts), 38)
})

test_that("Localization precision can be converted to covariance matrix", {
  PS <- lapply(PS, function(p) {
    p <- cbind(p, locprecz = p[, "locprec"])
  })
  C <- locprec2cov(PS, scale = FALSE)
  expect_equal(length(C), length(PS))
  expect_equal(length(dim(C[[1]])), 3)
})

test_that("Image can be generated from localizations by histogram method", {
  I <- points2img(points = as.data.frame(PS[[2]]), voxel.size = c(10, 10, 30), method = 'histogram')
  expect_equal(dim(I), c(30, 32, 29))
})

test_that("Image can be generated from localizations by photon method", {
  pts <- as.data.frame(cbind(PS[[2]], split(locs$phot, locs$site)[[2]]))
  colnames(pts)[ncol(pts)] <- "phot"
  suppressWarnings(I <- points2img(points = pts, voxel.size = c(10, 10, 30), method = 'photon', ncpu = 2))
  expect_equal(dim(I), c(30, 32, 29))
})

test_that("Circle Hough transform works", {
  I <- points2img(points = as.data.frame(PS[[2]]), voxel.size = c(10, 10, 30), method = 'histogram')
  img <- apply(I, c(1,2), sum) # 2D projection along z
  df <- circle_hough_transform(img, rmin = 5, rmax = 5, threshold = 0.4)
  expect_equal(df[, c("x", "y", "r")], data.frame(x = 16, y = 12, r = 5))
})

tests/testthat/test-registration.R
set.seed(20211123)

X <- read.csv(system.file("test_data", "parasaurolophusA.txt", package="LOMAR"), sep = "\t")
Y <- read.csv(system.file("test_data", "parasaurolophusB.txt", package="LOMAR"), sep = "\t")
Z <- read.csv(system.file("test_data", "parasaurolophusC.txt", package="LOMAR"), sep = "\t")

PS <- list(X, Y, Z)
C <- list()
for(i in 1:3) {
  cv <- diag(0.1, ncol(PS[[i]])) + jitter(0.01, amount = 0.01)
  cv <- replicate(nrow(PS[[i]]), cv)
  C[[i]] <- cv
}

test_that("Registration with cpd works", {
  trsf <- cpd(X, Y)
  expect_true(trsf$converged)
  expect_lt(trsf$sigma, 1e-5)
})

test_that("Registration with icp works", {
  trsf <- icp(X, Y, iterations = 20)
  expect_true(trsf$conv)
})

test_that("Joint registration of multiple point clouds works", {
  trsf <- jrmpc(PS, C = C, K = 100, maxIter = 20, tol = 0.01, model.selection = TRUE)
  expect_lte(trsf$iter, 20)
  expect_lte(trsf$conv, 0.02)
})

# test_that("Registration with wgmmreg works", {
#   trsf <- wgmmreg(PS[[1]], PS[[3]], C[[1]], C[[3]], maxIter = 20, tol = 1e-3)
#   expect_lte(trsf$c, 3e-3)
#   expect_true(trsf$converged)
# })


R/TDA.R
#' pssk
#' 
#' Compute the persistence scale-space kernel on persistence diagrams. 
#' Reference: Jan Reininghaus, Stefan Huber, Ulrich Bauer, and Roland Kwitt. A stable multi-scale kernel for topological machine learning. In Proceedings of the IEEE conference on computer vision and pattern recognition (CVPR), pages 4741–4748, 2015.
#'
#' @param Dg1 a persistence diagram as a n1 x 3 matrix where each row is a topological feature
#' and the columns are dimension, birth and death of the feature.
#' @param Dg2 another persistence diagram as a n2 x 3 matrix
#' @param sigma kernel bandwidth
#' @param dimensions vector of the dimensions of the topological features to consider, if NULL (default) use all available dimensions
#' @return kernel value
#' @examples 
#' D1 <- matrix(c(0,0,0,1,1,0,0,0,1.5, 3.5,2,2.5,3, 4, 6), ncol = 3, byrow = FALSE)
#' D2 <- matrix(c(0,0,1,1,0, 0, 1.2, 2, 1.4, 3.2,4.6,6.5), ncol = 3, byrow = FALSE)
#' K <- pssk(Dg1 = D1, Dg2 = D2, sigma = 1)
#' @export

pssk <- function(Dg1 = NULL, Dg2 = NULL, sigma = NULL, dimensions = NULL) {
  if (sigma <= 0) {
    stop("ERROR: Parameter sigma must be strictly positive.\n")
  }
  if(!is.null(dimensions)) {
    Dg1 <- Dg1[Dg1[,1] %in% dimensions,, drop = FALSE]
    Dg2 <- Dg2[Dg2[,1] %in% dimensions,, drop = FALSE]
  }
  dims <- unique(c(Dg1[,1], Dg2[,1]))
  K <- 0
  for(d in dims) {
    PD1 <- Dg1[Dg1[,1] == d, 2:3, drop = FALSE]
    PD2 <- Dg2[Dg2[,1] == d, 2:3, drop = FALSE]
    n1 <- nrow(PD1)
    n2 <- nrow(PD2)
    if(n1==0 || n2==0) { next }
    k <- 0
    # Note: nested loops 25% faster than using apply()
    for(i in 1:n1) {
      p <- PD1[i,]
      for(j in 1:n2) {
        q <- PD2[j,]
        qbar <- rev(q)
        d1 <- (p[1]-q[1])^2+(p[2]-q[2])^2
        d2 <- (p[1]-qbar[1])^2+(p[2]-qbar[2])^2
        k <- k + exp(-d1/(8*sigma)) -exp(-d2/(8*sigma))
      }
    }
    k <- k/(8*pi*sigma)
    K <- K + k
  }
  return(K)
}

#' sliced_Wd
#' 
#' Compute sliced Wasserstein distance or kernel. 
#' Reference: Mathieu Carriere, Marco Cuturi, and Steve Oudot. Sliced Wasserstein kernel for persistence diagrams. In Proceedings of the 34th International Conference on Machine Learning, volume 70 of Proceedings of Machine Learning Research, pages 664–673, 2017.
#'
#' @param Dg1 a persistence diagram as a n1 x 3 matrix where each row is a topological feature
#' and the columns are dimension, birth and death of the feature.
#' @param Dg2 another persistence diagram as a n2 x 3 matrix
#' @param M number of slices (default: 10)
#' @param sigma kernel bandwidth (default: 1)
#' @param dimensions  vector of the dimensions of the topological features to consider, if NULL (default) use all available dimensions
#' @param return.dist logical (default: FALSE). Whether to return the kernel or distance value.
#' @return kernel or distance value
#' @examples 
#' D1 <- matrix(c(0,0,0,1,1,0,0,0,1.5, 3.5,2,2.5,3, 4, 6), ncol = 3, byrow = FALSE)
#' D2 <- matrix(c(0,0,1,1,0, 0, 1.2, 2, 1.4, 3.2,4.6,6.5), ncol = 3, byrow = FALSE)
#' K <- sliced_Wd(Dg1 = D1, Dg2 = D2, M = 10, sigma = 1, return.dist = TRUE)
#' @export

sliced_Wd <- function(Dg1, Dg2, M = 10, sigma = 1, dimensions = NULL, return.dist = FALSE) {
 
  if(!is.null(dimensions)) {
    Dg1 <- Dg1[Dg1[,1] %in% dimensions,, drop = FALSE]
    Dg2 <- Dg2[Dg2[,1] %in% dimensions,, drop = FALSE]
  }
  dims <- unique(c(Dg1[,1], Dg2[,1]))
  SW <- 0
  for(d in dims) {
    PD1 <- Dg1[Dg1[,1] == d, 2:3, drop = FALSE]
    PD2 <- Dg2[Dg2[,1] == d, 2:3, drop = FALSE]
    n1 <- nrow(PD1)
    n2 <- nrow(PD2)
    if(n1==0 || n2==0) { next }
    # Project diagram points on the diagonal and add to the other diagram
    L <- c(2,2) # Vector representing the diagonal
    p.PD1 <- matrix(0, ncol = 2, nrow = n1)
    for(i in 1:n1) {
      p <- PD1[i,]
      s <- (L %*% p)/(L%*%L)
      p.PD1[i,] <- s[1,1] * L
    }
    p.PD2 <- matrix(0, ncol = 2, nrow = n2)
    for(i in 1:n2) {
      p <- PD2[i,]
      s <- (L %*% p)/(L%*%L)
      p.PD2[i,] <- s[1,1] * L
    }
    PD1 <- rbind(PD1, p.PD2)
    PD2 <- rbind(PD2, p.PD1)
    
    theta <- -pi/2
    s <- pi/M
    K <- 0
    V1 <- numeric(n1+n2)
    V2 <- numeric(n1+n2)
    for(i in 1:M) {
      L <- c(cos(theta), sin(theta))
      V1 <- apply(PD1, 1, function(x) { x %*% L })
      V1 <- sort(V1)
      V2 <- apply(PD2, 1, function(x) { x %*% L })
      V2 <- sort(V2)
      K <- K + s * (sum(abs(V1 - V2)))
      theta <- theta + s
    }
    K <- K/pi
    SW <- SW + K
  }
  if (!return.dist) {
    return(exp(-SW/sigma))
  } else {
    return(SW)
  }
}

#' get_persistence_diagrams
#' 
#' Compute persistence diagrams for a list of point sets.
#' By default, compute persistent homology from the Vietoris-Rips filtration.
#' If use.dtm is TRUE, compute instead the persistent homology of the sublevel
#' set of the distance to measure evaluated over a grid.
#' 
#' @importFrom foreach %dopar%
#' @param point.sets list of point sets, each as a data frame with columns x,y,z
#' @param maxdimension maximum dimension of the homological features to be computed
#' @param maxscale limit of the Vietoris-Rips filtration
#' @param use.dtm logical (default: FALSE), whether to use the distance to measure function
#' @param m0 parameter for the dtm function
#' @param grid.by vector of space between points of the grid for the dtm function along each dimension
#' @param ncpu number of parallel threads to use for computation
#' @return a list of persistence diagrams as n x 3 matrices. Each row is a topological feature
#'  and the columns are dimension, birth and death of the feature.
#' @examples 
#' PS <- list(data.frame(x = c(2.4,-6.9,4.6,-0.7,-3.3,-4.9,-3.5,-3.5,4.2,-7),
#'                       y = c(5.7,1.9,4.8,3.4,-3,-2.1,7.2,1.8,6.1,-1.6),
#'                       z = c(2.7,-0.1,-0.7,-0.6,0.4,-1.5,-0.6,-0.9,2.2,0.7)),
#'            data.frame(x = c(0,0,3.1,-5.6,-5,-7.4,-0.7,-7.7,-6.7,4,4.2,0.2,5.8,3.9,3.9),
#'                       y = c(6.3,-6.1,-3.5,4.6,-4.1,0.3,8.8,-2.3,2.9,3.7,-1.4,-3.9,5.5,-1.2,-6.7),
#'                       z = c(-1.5,1.7,-0.4,-1.4,1.8,1.7,-0.9,-1.8,-0.5,1.7,1.3,0.5,-1.4,1.6,-0.1)))
#' Diags <- get_persistence_diagrams(point.sets = PS, maxdimension = 1, maxscale = 5, ncpu = 1)
#' @export

get_persistence_diagrams <- function(point.sets = NULL, 
                                     maxdimension = NULL, 
                                     maxscale = NULL, 
                                     use.dtm = FALSE,
                                     m0 = NULL,
                                     grid.by = NULL,
                                     ncpu = 1) {
  Diag <- list()
  i <- NULL
  # Set up cluster for parallelization
  cluster <- parallel::makeCluster(ncpu)
  doParallel::registerDoParallel(cluster)
  n <- length(point.sets)
  if(!use.dtm) {
    Diag <- foreach::foreach(i = c(1:n), .packages = "LOMAR") %dopar% {
      ripsDiag(X = point.sets[[i]],
               maxdimension,
               maxscale,
               printProgress = FALSE)[["diagram"]]
    }
  } else {
    point.sets <- lapply(point.sets, function(x) {scale(x, center = TRUE, scale = FALSE)})
    min.dims <- apply(sapply(point.sets, function(x) {apply(x, 2, min)}), 1, min)
    max.dims <- apply(sapply(point.sets, function(x) {apply(x, 2, max)}), 1, max)
    lims <- NULL
    for(l in 1:length(max.dims)) {
      lims <- cbind(lims, c(min.dims[l], max.dims[l]))
    }
    Diag <- foreach::foreach(i = c(1:n), .packages = "LOMAR") %dopar% {
      gridDiag(X = point.sets[[i]], 
               FUN = dtm, 
               m0 = m0, 
               lim = lims,
               by = grid.by, 
               sublevel = TRUE,
               maxdimension = maxdimension,
               printProgress = FALSE )$diagram
    }
  }
  on.exit(parallel::stopCluster(cluster))
  return(Diag)
}

#' get_kernel_matrix
#' 
#' Compute kernel/distance matrix between persistence diagrams.
#' 
#' @importFrom foreach %dopar%
#' @importFrom foreach %:%
#' @param Diag list of persistence diagrams as n x 3 matrices
#' @param method which kernel or distance to compute. One of sWd (for sliced Wasserstein kernel) or pssk (for the persistence scale-space kernel)
#' @param sigma kernel bandwidth
#' @param return.dist logical (default: FALSE) for method sWd, whether to return the sliced Wasserstein distance matrix instead of the kernel.
#' @param M number of slices for the sliced Wasserstein kernel
#' @param dimensions vector of the dimensions of the topological features to consider, if NULL (default) use all available dimensions
#' @param ncpu number of parallel threads to use for computation
#' @return a matrix
#' @examples 
#' PS <- list(data.frame(x = c(2.4,-6.9,4.6,-0.7,-3.3,-4.9,-3.5,-3.5,4.2,-7),
#'                       y = c(5.7,1.9,4.8,3.4,-3,-2.1,7.2,1.8,6.1,-1.6),
#'                       z = c(2.7,-0.1,-0.7,-0.6,0.4,-1.5,-0.6,-0.9,2.2,0.7)),
#'            data.frame(x = c(0,0,3.1,-5.6,-5,-7.4,-0.7,-7.7,-6.7,4,4.2,0.2,5.8,3.9,3.9),
#'                       y = c(6.3,-6.1,-3.5,4.6,-4.1,0.3,8.8,-2.3,2.9,3.7,-1.4,-3.9,5.5,-1.2,-6.7),
#'                       z = c(-1.5,1.7,-0.4,-1.4,1.8,1.7,-0.9,-1.8,-0.5,1.7,1.3,0.5,-1.4,1.6,-0.1)),
#'            data.frame(x = c(-9.8,-5.2,12.5,2.5,4.5,1.3,-0.2,0.4,9.3,-1.4,0.5,-1.1,-7.7),
#'                       y = c(-4.2,1.5,-0.5,12,-3,-7.2,10.9,6.7,-1.3,10,6.7,-6.2,2.9),
#'                       z = c(3.4,-3.8,-1.4,1.8,3.5,2.5,2.6,-4.8,-3.8,3.9,4.1,-3.6,-4)))
#' Dgs <- get_persistence_diagrams(point.sets = PS, maxdimension = 1, maxscale = 5, ncpu = 1)
#' K <- get_kernel_matrix(Diag = Dgs, method = 'sWd', dimensions = c(0,1), M = 10, sigma = 5)
#' @export

get_kernel_matrix <- function(Diag = NULL, method = c("sWd", "pssk"), dimensions = NULL, return.dist = FALSE, M = NULL, sigma = NULL, ncpu = 1) {
  n <- length(Diag)
  K <- matrix(NA, nrow = n, ncol = n)
  i <- j <- NULL
  cluster <- parallel::makeCluster(ncpu)
  doParallel::registerDoParallel(cluster)
  if(method == "sWd") {
    # Compute pairwise sliced Wasserstein distances
    K <- foreach::foreach(j = c(1:n), .combine='cbind') %:%
      foreach::foreach(i = c(1:n), .combine = 'c') %dopar% {
        Di <- Diag[[i]]
        Dj <- Diag[[j]]
        sliced_Wd(Di, Dj, M, sigma, dimensions, return.dist)
      }
  } else if (method == "pssk") {
    K <- foreach::foreach(j = c(1:n), .combine='cbind') %:%
      foreach::foreach(i = c(1:n), .combine = 'c') %dopar% {
        Di <- Diag[[i]]
        Dj <- Diag[[j]]
        pssk(Di, Dj, sigma, dimensions)
      }
  } else {
    stop(paste0("Unknown kernel method: ", method,".\n"))
  }
  on.exit(parallel::stopCluster(cluster))
  return(K)  
}



### Functions copied from package TDA ###

ripsDiag <- function(X, maxdimension, maxscale, location = FALSE, printProgress = FALSE) {
    
    if (!is.numeric(X) && !is.data.frame(X)) {
      stop("X should be a matrix of coordinates")
    }
    if (!is.numeric(maxdimension) ||
        length(maxdimension) != 1 || maxdimension < 0) {
      stop("maxdimension should be a nonnegative integer")
    }
    if (!is.numeric(maxscale) || length(maxscale) != 1) {
      stop("maxscale should be a number")
    }
    
    X <- as.matrix(X)
    
    max_num_pairs <- 5000  # to be added as an option
    
    ripsOut <- RipsDiag(X = X, maxdimension = maxdimension,
                        maxscale = maxscale, dist = 'euclidean', libraryFiltration = 'Dionysus',
                        libraryDiag = 'Dyonisus', location = location,
                        printProgress = printProgress)
    
    if (location == TRUE) {
      BirthLocation <- X[ripsOut[[2]][, 1], ]
      DeathLocation <- X[ripsOut[[2]][, 2], ]
    }
    
    Diag <- ripsOut[[1]]
    if (NROW(Diag) > 0) {
      ## change Inf values to maxscale
      Diag[which(Diag[, 3] == Inf), 3] <- maxscale  
    }
    
    colnames(Diag) <- c("dimension", "Birth", "Death")
    class(Diag) <- "diagram"
    attributes(Diag)[["maxdimension"]] <- max(Diag[, 1])
    attributes(Diag)[["scale"]] <- c(0, maxscale)
    attributes(Diag)[["call"]] <- match.call()
    
    out <- list("diagram" = Diag)
  
    return (out)
  }


gridDiag <- function(X = NULL, FUN = NULL, lim = NULL, by = NULL, FUNvalues = NULL,
           maxdimension = max(NCOL(X), length(dim(FUNvalues))) - 1,
           sublevel = TRUE, location = FALSE,
           printProgress = FALSE, diagLimit = NULL, ...) {
    
    if (!xor(is.null(X) || is.null(FUN), is.null(FUNvalues))) {
      stop("either values of X, FUN should be set, or a value of FUNvalues should be set, but not both")
    }
    if (!is.null(X) && !is.null(FUN)) {
      if (!is.numeric(X) && !is.data.frame(X)) {
        stop("X should be a matrix of coordinates")
      }
      if (!is.function(FUN)) {
        stop("FUN should be a function")
      }
      if (is.null(lim) || is.null(by)) {
        stop("When X and FUN are set, lim and by should also be set")
      }
    }
    if (!is.null(lim) && !is.null(by)) {
      if (!is.numeric(lim) || length(lim) %% 2 != 0) {
        stop("lim should be either a numeric matrix or a numeric vector of even length")
      }
      if (!is.numeric(by) || any(by <= 0)) {
        stop("by should be positive")
      }
    }
    if (!is.null(X) && !is.null(FUN) && !is.null(lim) && !is.null(by)) {
      if (2 * NCOL(X) != length(lim)) {
        stop("dimension of X does not match with lim")
      }
      if (length(by) != 1 && length(by) != NCOL(X)) {
        stop("by should be either a number or a vector of length equals dimension of grid")
      }
    }
    if (!is.null(FUNvalues)) {
      if (!is.numeric(FUNvalues) && !is.data.frame(FUNvalues)) {
        stop("FUNvalues should be an array")
      }
    }
    if (!is.numeric(maxdimension) || length(maxdimension) != 1 ||
        maxdimension < 0) {
      stop("maxdimnsion should be a nonnegative integer")
    }
    if (!is.logical(sublevel)) {
      stop("sublevel should be logical")
    }
    if (!is.logical(location)) {
      stop("location should be logical")
    }
    if(location == TRUE && (is.null(lim) || is.null(by))) {
      stop("If location is TRUE, both lim and by should be set")
    }
    if (!is.logical(printProgress)) {
      stop("printProgress should be logical")
    }
    if ((!is.numeric(diagLimit) || length(diagLimit) != 1) &&
        !is.null(diagLimit)) {
      stop("diagLimit should be a positive number")
    }
    
    if (!is.null(X)) {
      X <- as.matrix(X)
    }
    if (!is.null(lim) && !is.null(by)) {
      Grid <- gridBy(lim = lim, by = by)
    }
    if (is.null(FUNvalues)) {
      FUNvalues <- FUN(X, Grid[["grid"]], ...)
      gridDim <- Grid[["dim"]]
    } else {
      if (is.data.frame(FUNvalues)) {
        FUNvalues <- as.matrix(FUNvalues)
      } else {
        FUNvalues <- as.array(FUNvalues)
      }
      gridDim <- dim(FUNvalues)
    }
    
    if (!is.null(lim) && !is.null(by) &&
        length(Grid[["dim"]]) != length(gridDim)) {
      stop("dimension of FUNvalues does not match with lim and by")
    }
    
    maxdimension <- min(maxdimension, length(gridDim) - 1)
    if (sublevel == FALSE) {
      FUNvalues <- -FUNvalues
    }
    
    # compute persistence diagram of function values over a grid
    if (length(gridDim) <= 3) {
      gridOut <- GridDiag(FUNvalues = FUNvalues, gridDim = as.integer(gridDim),
        maxdimension = as.integer(maxdimension), decomposition = "5tetrahedra",
        library = 'Dyonisus', location = location, printProgress = printProgress)
    } else {
      gridOut <- GridDiag(FUNvalues = FUNvalues, gridDim = as.integer(gridDim),
        maxdimension = as.integer(maxdimension), decomposition = "barycenter",
        library = 'Dyonisus', location = location, printProgress = printProgress)
    }
    
    if (location == TRUE) {
      BirthLocation <- Grid[["grid"]][gridOut[[2]][, 1], ]
      DeathLocation <- Grid[["grid"]][gridOut[[2]][, 2], ]
    }
    
    Diag <- gridOut[[1]]
    if (NROW(Diag) > 0) {
      Diag[1, 3] <- ifelse(is.null(diagLimit), max(FUNvalues), diagLimit) 
    }
    if (sublevel == FALSE) {
      colnames(Diag) <- c("dimension", "Death", "Birth")
      Diag[, 2:3] <- -Diag[, 3:2]
    } else {
      colnames(Diag) <- c("dimension", "Birth", "Death")
    }
    
    class(Diag) <- "diagram"
    attributes(Diag)[["maxdimension"]] <- max(Diag[, 1])
    nonInf <- which(Diag[, 2] != Inf & Diag[, 3] != Inf)
    attributes(Diag)[["scale"]] <-
      c(min(Diag[nonInf, 2:3]), max(Diag[nonInf, 2:3]))
    attributes(Diag)[["call"]] <- match.call()
    
    out <- list("diagram" = Diag)
    
    return (out)
  }

gridBy <- function(lim = c(0, 1), by = (lim[2] - lim[1]) / 10) {
  gridlist <- list()
  for (idx in 1:(length(lim)/2)) {
    if (length(by) == 1) {
      gridlist[[idx]] <- seq(lim[2 * idx - 1], lim[2 * idx], by = by)
    } else {
      gridlist[[idx]] <- seq(lim[2 * idx - 1], lim[2 * idx], by = by[idx])
    }
  }
  grid <- as.matrix(expand.grid(gridlist))
  colnames(grid) <- NULL
  dim <- sapply(gridlist, length)
  out <- list("dim" = dim, "lim" = lim, "grid" = grid)
  return(out)
}

dtm <- function(X, Grid, m0, r = 2, weight = 1) {

  if (!is.numeric(X) && !is.data.frame(X)) {
    stop("X should be a matrix of coordinates")
  }
  if (!is.numeric(Grid) && !is.data.frame(Grid)) {
    stop("Grid should be a matrix of coordinates")
  }
  if (NCOL(X) != NCOL(Grid)) {
    stop("dimensions of X and Grid do not match")
  }
  if (!is.numeric(m0) || length(m0) != 1 || m0 < 0 || m0 > 1) {
    stop("m0 should be a number between 0 and 1")
  }
  if (!is.numeric(r) || length(r) != 1 || r < 1) {
    stop("r should be a number greater than or equal to 1")
  }
  if (!is.numeric(weight) || 
      (length(weight) != 1 && length(weight) != NROW(X))) {
    stop("weight should be either a number or a vector of length equals the number of sample")
  }

  # without weight
  if (length(weight) == 1) {
    X <- as.matrix(X)
    weightBound <- m0 * NROW(X)
    knnDistance <- FNN::knnx.dist(
		data = X, query = as.matrix(Grid), k = ceiling(weightBound),
		algorithm = c("kd_tree"))
    return (Dtm(knnDistance = knnDistance, weightBound = weightBound, r = r))

  # with weight
  } else {
    X0 <- as.matrix(X[weight != 0, , drop = FALSE]) 
    weight0 <- weight[weight != 0]
    weight0sort <- sort(weight0)
    weightBound <- m0 * sum(weight0)
    weightSumTemp <- 0
    for (k0 in seq(along = weight0)) {
      weightSumTemp <- weightSumTemp + weight0sort[k0]
      if (weightSumTemp >= weightBound) {
        break
      }
    }
    knnDistanceIndex <- FNN::get.knnx(
	    data = X0, query = as.matrix(Grid), k = k0, algorithm = c("kd_tree"))
    return (DtmWeight(
	    knnDistance = knnDistanceIndex[["nn.dist"]], weightBound = weightBound,
		r = r, knnIndex = knnDistanceIndex[["nn.index"]], weight = weight0))
  }
}
